src/subject_name/mod.rs

```diff
@@ -13,13 +13,11 @@
 // OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 
 mod dns_name;
-#[cfg(feature = "alloc")]
-pub(crate) use dns_name::GeneralDnsNameRef;
 pub use dns_name::{DnsNameRef, InvalidDnsNameError};
 
 /// Requires the `alloc` feature.
 #[cfg(feature = "alloc")]
-pub use dns_name::DnsName;
+pub use dns_name::{DnsName, GeneralDnsNameRef, WildcardDnsNameRef};
```

Comment:
What is the perceived use case for making these public?

Reply:
Two separate people have asked for it now in #178. It seemed like there was value in the downstream code knowing whether a name contained a wildcard label or not without having to reconstitute the determination webpki uses for that.

Reply:
Not sure I fully understand #50 -- it appears we have some basic integration test coverage of `dns_names()` that covers normal DNS and wildcard names. That shows the feature is usable outside the crate. What did I miss?

(I think the tests would benefit from coverage that `dns_names()` _doesn't_ yield IP address names, but that's unrelated.)

Reply:
> That shows the feature is usable outside the crate.

You have to do a bit of a dance to collect the general name iterator to a collection of `&str` which loses the context from webpki about whether a given DNS name is a wildcard DNS name or not.

Reply:
> You have to do a bit of a dance to collect the general name iterator to a collection of &str

If that's sufficient for consumers we could also just rework `dns_names` to return &str's :shrug:

Reply:
@hawkw Can you speak to the use-case [you had in mind](https://github.com/rustls/webpki/pull/50#issuecomment-1703156932) when requesting `GeneralDnsNameRef` be exported?

Reply:
> @hawkw Can you speak to the use-case [you had in mind](https://github.com/rustls/webpki/pull/50#issuecomment-1703156932) when requesting `GeneralDnsNameRef` be exported?

Turns out that the `AsRef<str>` implementation was sufficient for my use-case. It just wasn't obvious from the documentation that you could just `as_ref()` the items of the iterator and get an `&str` --- so when looking at the docs, I thought it was an iterator over types I couldn't do anything with in downstream code.

It might be more useful to just return an iterator of `&str`s...

Reply:
I'll give this some thought and rework. I'm still thinking there's value in letting consumers differentiate wildcard names but I think the majority will just want `&str`. In either case I agree w/ Djc that the type names could use adjustment if we're going to put them in front of users.

Reply:
I'm certainly not opposed to an API where consumers can match on a `GeneralDnsNameRef` to distinguish between wildcards and other names --- I think that's potentially useful, too! IMO, we should either make the type public *or* make the iterator be over `&str`s, because a private type which can only be used as `impl AsRef<str>` is just unnecessary ceremony for no reason. Publicly exporting `GeneralDnsNameRef` would solve that problem, as would avoiding it in the API.

Reply:
I switched to providing an iterator of `&str` and removed the exports of the `GeneralDnsNameRef` and `WildcardDnsNameRef` types.